### PR TITLE
Replace next image optimization with drupal image styles

### PR DESCRIPTION
**Pull request by pookmish**
# READY FOR REVIEW

# Summary
- Provide a custom next image loader that uses drupal image styles.

# Steps to Test

1. review the approach.

### CODE REVIEW

**Comment by pookmish** [comment]
Thinking about this, we loose webp support. Not sure if that's a concern, but it is something to note

File: next.config.js [hunk 1]
```diff
@@ -9,27 +9,32 @@ module.exports = {
         hostname: process.env.NEXT_IMAGE_DOMAIN,
       },
     ],
+    deviceSizes: [
```
Comment (sherakama) [inline next.config.js:12]: Can you tell me why you chose these breakpoints?

Comment (pookmish) [reply next.config.js:12]: Mostly arbitrary. I took the default values and commented out the ones I didn't want. I tried to get something close-ish to the breakpoints.

File: lib/libguides.tsx [hunk 2]
```diff
@@ -21,6 +21,7 @@ const fetchLibGuides = async ({accountId, subjectId}: { accountId?: number, subj
     }
 
     const guidesConfig = {
+      next: {revalidate: 1},
```
Comment (sherakama) [inline lib/libguides.tsx:24]: What is this doing?

File: components/patterns/elements/image-loader.tsx [hunk 3]
```diff
@@ -0,0 +1,13 @@
+const ImageLoader = ({ src, width }) => {
+  const imageStyles = [
+    {w: 128, s: 'breakpoint_sm_2x'},
+    {w: 384, s: 'breakpoint_md_2x'},
+    {w: 750, s: 'breakpoint_lg_2x'},
+    {w: 1200, s: 'breakpoint_xl_2x'},
+    {w: 3840, s: 'breakpoint_2xl_2x'},
+  ];
+  const style = imageStyles.find(s => s.w === width)?.s || 'breakpoint_xl_2x';
```
Comment (sherakama) [inline components/patterns/elements/image-loader.tsx:9]: Writing this out in English so my brain is happy.

For all images, that are not local or relative, try and match an image style to an image width, if no matching image width to the array of imageStyles, default to the biggest image style.

File: components/patterns/elements/image-loader.tsx [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  // Switch the image style for the desired width. Add the width parameter just to prevent console logs.
```
Comment (sherakama) [inline components/patterns/elements/image-loader.tsx:10]: Eh, what console log?

File: next.config.js (hunk 1 above)
Comment (sherakama) [reply next.config.js:12]: Why not align them with the breakpoints of the design system?

File: next.config.js [hunk 5]
```diff
@@ -9,27 +9,32 @@ module.exports = {
...
+    imageSizes: [
+      // 16,
+      // 32,
+      // 48,
+      // 64,
+      // 96,
+      128,
+      // 256,
+      384
+    ],
```
Comment (sherakama) [inline next.config.js:31]: What unit are these in? I'm not fully grokking the documentation 

https://nextjs.org/docs/app/api-reference/components/image#imagesizes

https://nextjs.org/docs/app/api-reference/components/image#sizes

Do you have an example I can follow?

File: next.config.js [hunk 6]
```diff
@@ -9,27 +9,32 @@ module.exports = {
...
-        {
-          source: '/_next/image',
-          destination: '/_next/image?url=/no-image.png',
-          has: [{type: 'query', key: 'url', value: (`htt[p|ps]://${process.env.NEXT_IMAGE_DOMAIN}.*`)}],
-          missing: [{type: 'query', key: 'url', value: '(.*itok=([\\w|-]+))'}]
-        },
-        {
-          source: '/_next/image',
-          destination: '/_next/image?url=:url',
-          has: [{type: 'query', key: 'url', value: '(?<url>.*[jpg|png|jpeg|gif]\?itok=([\\w|-]+)).*'}]
```
Comment (sherakama) [inline next.config.js:28]: Is this not working due to the bot trolling URLs that look legit?

**Comment by sherakama** [comment]
Do you have to do anything to support pixel density (1x and 2x)?

File: components/patterns/elements/image-loader.tsx (hunk 3 above)
Comment (sherakama) [reply components/patterns/elements/image-loader.tsx:9]: Example: https://github.com/SU-SWS/sulgryphon-nextjs/blob/1.x/components/paragraph/stanford-card.tsx#L43

That wouldn't get all the src set values (128, 384, 750, 1200, 3840) because it doesn't set the `sizes` key? It would just render as the one size?

**Comment by sherakama** [comment]
Oh and one last thought, what about cache control? https://nextjs.org/docs/app/api-reference/components/image#caching-behavior

File: next.config.js (hunk 1 above)
Comment (pookmish) [reply next.config.js:12]: meh, it's not really necessary IMO. Images don't need to break down at the same breakpoint. Just as long as the image is larger than the area it's in. Almost all images use a css cover property anyways

File: next.config.js (hunk 5 above)
Comment (pookmish) [reply next.config.js:31]: https://html.com/attributes/img-srcset/#Specifying_Image_Width

So it tells the browser the size of the image... which isn't really accurate, but it's something
```
<img alt="" loading="lazy" decoding="async" data-nimg="fill" class="su-object-cover su-object-center" style="position: absolute; height: 100%; width: 100%; inset: 0px; color: transparent;" sizes="100vw" srcset="https://example.com/breakpoint_lg_2x/image.jpg 750w, https://example.com/breakpoint_xl_2x/image.jpg 1200w, https://example.com/breakpoint_2xl_2x/image.jpg 3840w" src="https://example.com/breakpoint_2xl_2x/image.jpg">
```

**Comment by pookmish** [comment]
> Do you have to do anything to support pixel density (1x and 2x)?

Nothing provided to handle that

> Oh and one last thought, what about cache control?

The image src urls are changed to target the Drupal site directly. So the cacheing would be on the Acquia side.

A problem I found recently though is this implementation requires the `allow_insecure_derivatives` enabled because the `itok` for each image style would be different.

File: lib/libguides.tsx (hunk 2 above)
Comment (sherakama) [reply lib/libguides.tsx:24]: Aye, but why is it in this PR?

Comment (pookmish) [reply lib/libguides.tsx:24]: :shrug: it was in my local code, so i just committed it with the stuff.

File: components/patterns/elements/image-loader.tsx (hunk 3 above)
Comment (sherakama) [reply components/patterns/elements/image-loader.tsx:10]: Gotcha thanks.

So the convention and recommendation would be to supply a sizes property wherever images are being used, otherwise the largest image will be used?

File: components/patterns/elements/image-loader.tsx (hunk 4 above)
Comment (sherakama) [reply components/patterns/elements/image-loader.tsx:11]: Huh, that is rather presumptive of Next and 3rd party image optimization services.

File: components/patterns/elements/image-loader.tsx [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    {w: 384, s: 'breakpoint_md_2x'},
+    {w: 400, s: 'breakpoint_md_2x'},
+    {w: 750, s: 'breakpoint_lg_2x'},
+    {w: 1200, s: 'breakpoint_xl_2x'},
+    {w: 3840, s: 'breakpoint_2xl_2x'},
+  ];
+  const style = imageStyles.find(s => s.w === width)?.s || 'breakpoint_xl_2x';
+  // Switch the image style for the desired width. Add the width parameter just to prevent console logs.
+  // https://nextjs.org/docs/messages/next-image-missing-loader-width
+  return src.toString().replace(/styles\/\w+\//, `styles/${style}/`) + `&w=${width}`;
```
Comment (sherakama) [inline components/patterns/elements/image-loader.tsx:13]: However unlikely, in theory, the src string could already have a `w` parameter. It would be safer to put the URL through  `new URL` and then set the searchParams instead of concatenating the string.

eg: https://nextjs.org/docs/app/api-reference/next-config-js/images#fastly

File: components/patterns/elements/image-loader.tsx (hunk 3 above)
Comment (pookmish) [reply components/patterns/elements/image-loader.tsx:10]: no no... It's recommended to use the default values. Maybe unless you are displaying a small image and don't need the large versions. If you don't pass a `sizes` property when using an `<Image>` it will make a responsive image using the sizes defined in the next.config.js which should be the most ideal in most situations I think.

File: components/patterns/elements/image-loader.tsx (hunk 7 above)
Comment (pookmish) [reply components/patterns/elements/image-loader.tsx:13]: I thought about that too. It would be more bullet proof that way, but since we're only ever pulling images from Drupal, there shouldn't ever be that param.

File: next.config.js (hunk 1 above)
Comment (sherakama) [reply next.config.js:12]: This seems like a thing to ask what others on the team are doing and how they set up their responsive images. It is in the 'doesn't really matter' category but this seems like a good opportunity to engage and align with others.

**Comment by pookmish** [comment]
FYI, I'm still not 100% convinced myself of this approach since it requires insecure derivatives in the Drupal side. What are your thoughts on that

File: lib/libguides.tsx (hunk 2 above)
Comment (sherakama) [reply lib/libguides.tsx:24]: Ok, if you want it in then I guess it stays in. 

It would be great if you could write out an RFC-style document that details how the ISR invalidation is intended to work so that I could have something to compare this to. That would also be helpful documentation for the SUP project.

File: components/patterns/elements/image-loader.tsx (hunk 3 above)
Comment (sherakama) [reply components/patterns/elements/image-loader.tsx:10]: Gotcha, ok, this info is something that @jenbreese will find helpful/useful when she gets going on the SUP project.

**Comment by sherakama** [comment]
> FYI, I'm still not 100% convinced myself of this approach since it requires insecure derivatives in the Drupal side. What are your thoughts on that

By skipping the Next/Vercel optimizations and going straight to Drupal we are removing the DoS flood protection on Drupal and limiting the styles/sizes that can be produced to the image styles available in the Drupal install. Overall, I don't think we've had many issues with DoS attacks on our image derivatives and we have better logging/reporting/blocking on Acquia than we do currently on Vercel so this seems like a valid trade.

However, if we are going to remove the secure token, why not just re-write the image paths, like you had before, to remove any and all URL parameters and force the image source to always be the original file? Wouldn't that get around the URL parameter 'source' issue on Vercel and allow this site to continue to use Next's image optimizations and Vercel's Edge network caching?

**Comment by pookmish** [comment]
> However, if we are going to remove the secure token, why not just re-write the image paths, like you had before, to remove any and all URL parameters and force the image source to always be the original file? Wouldn't that get around the URL parameter 'source' issue on Vercel and allow this site to continue to use Next's image optimizations and Vercel's Edge network caching?

From what it seems, it doesn't matter if we rewrite the urls. Vercel still treats each request as a unique value. I tried to implement a middleware for the `_next/image` path and it didn't seem to work. I then tried to use vercel.json file to redirect paths instead of rewrite them. That  also didn't seem to work as desired. I'm not sure if our Akamai service comes with an image optimizer as [mentioned here](https://nextjs.org/docs/app/api-reference/next-config-js/images#akamai), so the Drupal image styles seem to be the only option to fix this.

**Comment by sherakama** [comment]
> > However, if we are going to remove the secure token, why not just re-write the image paths, like you had before, to remove any and all URL parameters and force the image source to always be the original file? Wouldn't that get around the URL parameter 'source' issue on Vercel and allow this site to continue to use Next's image optimizations and Vercel's Edge network caching?
> 
> From what it seems, it doesn't matter if we rewrite the urls. Vercel still treats each request as a unique value. I tried to implement a middleware for the `_next/image` path and it didn't seem to work. I then tried to use vercel.json file to redirect paths instead of rewrite them. That also didn't seem to work as desired. I'm not sure if our Akamai service comes with an image optimizer as [mentioned here](https://nextjs.org/docs/app/api-reference/next-config-js/images#akamai), so the Drupal image styles seem to be the only option to fix this.

Remote image invalidation is supported by a cache query param: https://vercel.com/docs/image-optimization#remote-images
> Remote image cache invalidation: Adding a query string to the image URL (e.g., ?new or ?v=2) will serve a new image to visitors. Alternatively, you can configure the cache to expire more frequently in order to serve new images.

But I am surprised you cannot re-write the URL to drop the extra params before the route is hit or provide the middleware to validate that the request came from a legit page render by using a CSRF token or something. 

I submitted a support request with Vercel
https://vercel.com/teams/stanford-libraries/support/cases/00179738

One thing we may want to do is forego fetching from the image styles directory at all and always pull the original source image from sites/default/files. Optimize on Vercel for different sizes of the one source file instead of optimizing on the already optimized file.

**Comment by sherakama** [comment]
Reply from Vercel support

> The best way to filter out these cache busting requests would be to implement Edge Middleware that filters out query parameters: Edge Middleware Overview
> 
> We have a pre-existing template on this which you might find helpful: Filtering Query Parameters
> 
> This middleware allows you to specify a list of allowed query parameters (`allowedParams`), and any query parameter not included in this list is removed from the URL, effectively filtering out unwanted or potentially harmful parameters.
> 
> If the offender is consistently making these requests from a specific IP address, you can also leverage our middleware template on IP Blocking with Upstash or IP Blocking with DataDome.
> 
> You can also introduce User-Agent Based Rendering or API Rate Limiting by IP and API Keys with Upstash depending on your application's needs.

**Comment by pookmish** [comment]
I kinda had a thought about the whole itok thing... what if we removed the use of the image styles like you mentioned. then in the components, it will be fetching the original image which doesn't need an itok at all. This would eliminate the sql injection getting appended to the parameter. 

However, I don't think it solves the whole problem, because based on the documentation, a bot can just add it to the `url` to generate another image. `/_next/image?url=http....jpg'inject=something'&w=1200`.

**Comment by sherakama** [comment]
> I kinda had a thought about the whole itok thing... what if we removed the use of the image styles like you mentioned. then in the components, it will be fetching the original image which doesn't need an itok at all. This would eliminate the sql injection getting appended to the parameter.
> 
> However, I don't think it solves the whole problem, because based on the documentation, a bot can just add it to the `url` to generate another image. `/_next/image?url=http....jpg'inject=something'&w=1200`.

That sounds like an improvement to me.

As far as parameters go, I think your rewrite is helping and can continue to help. If you can adjust it to allow only the `w` parameter and only contain a max of 4 integers, I think you'll reduce the attack surface substantially. Yes, bots could send arbitrary integers directly to that URL param and at that point, we'd need to block the bot. 

With the metrics add on enabled I've been trying to pull the referrer or source for the image endpoint. I can see that on December 1 between 12:00pm - 2:00pm there was a couple of IPs (from the east coast) hitting that endpoint more than others. 

https://vercel.com/dashboard/stanford-libraries/monitoring?visualize=sum%28requests%29&groupBy=public_ip&title=Template%3A+Requests+by+IP+Address&startTime=1701460800000&deltaTime=7140000&g=hour

Filtering by bots hasn't shown me anything to go on

https://vercel.com/dashboard/stanford-libraries/monitoring?visualize=sum%28requests%29&groupBy=bot_name&title=Template%3A+Requests+by+Bot%2FCrawler&startTime=1701460800000&deltaTime=7140000&g=hour

And user agents for the period

https://vercel.com/dashboard/stanford-libraries/monitoring?visualize=sum%28requests%29&groupBy=user_agent&title=Template%3A+Requests+by+User+Agent&startTime=1701460800000&deltaTime=7140000&g=hour

Perhaps [Faraday](https://faradaysec.com/) but that doesn't account for enough of the requests. 

One more query

https://vercel.com/dashboard/stanford-libraries/monitoring?visualize=sum%28requests%29&groupBy=source_path%2Chttp_referer%2Cuser_agent&title=Template%3A+Execution+by+Host%2C+Path&where=startsWith%28source_path%2C+%27%2Fsites%2Flibrary%2Ffiles%27%29&h=&md=25&startTime=1701460800000&deltaTime=7140000&g=hour

So TLDR; Nothing is standing out for me right now on who or what to block.

**Comment by pookmish** [comment]
Closing this since we seem to be doing good since the issue.
